File: BASEJAVASCRIPT/backend/src/services/orderPickup.js
'use strict';

/** Filas nativas + pickup (port de native_queues.py / order_pickup.py) */

const { normalizeRole, ROLE_ADMIN, ROLE_TECNICO, ROLE_EXPEDICAO } = require('./operatorRoles');
const db = require('../db');

const PERSONAL_QUEUE_PREFIX = 'Fila · ';
const PERSONAL_STATUS_ID_BASE = 900000;

const PICKABLE_BY_ROLE = {
  [ROLE_TECNICO]: [
    'Novos pedidos',
    'Notebook - Geral',
    'Computadores - Geral',
    'Fila Técnico',
    'Pedidos Agendados',
    'Para Enviar Amanhã',
  ],
  [ROLE_EXPEDICAO]: ['Em Separação - Geral', 'Separação Finalizada'],
  [ROLE_ADMIN]: [
    'Novos pedidos',
    'Notebook - Geral',
    'Computadores - Geral',
    'Fila Técnico',
    'Em Separação - Geral',
    'Pedidos Agendados',
    'Para Enviar Amanhã',
    'Separação Finalizada',
  ],
};

const DEFAULT_SEND_BY_ROLE = {
  [ROLE_TECNICO]: 'Em Separação - Geral',
  [ROLE_EXPEDICAO]: 'Pronto P/ Envio',
  [ROLE_ADMIN]: 'Em Separação - Geral',
};

function personalQueueName(operatorName) {
  return `${PERSONAL_QUEUE_PREFIX}${(operatorName || 'Operador').trim() || 'Operador'}`;
}

function personalStatusId(operatorId) {
  return PERSONAL_STATUS_ID_BASE + Number(operatorId);
}

function isPersonalStatusName(name) {
  const n = (name || '').trim();
  return n.startsWith(PERSONAL_QUEUE_PREFIX) || n.toLowerCase().startsWith('fila ·');
}

function isPersonalStatusId(statusId) {
  try {
    const sid = Number(statusId || 0);
    return sid >= PERSONAL_STATUS_ID_BASE;
  } catch {
    return false;
  }
}

function pickableQueueNames(role) {
  const canon = normalizeRole(role);
  return new Set(PICKABLE_BY_ROLE[canon] || PICKABLE_BY_ROLE[ROLE_TECNICO]);
}

function defaultSendQueueName(role) {
  return DEFAULT_SEND_BY_ROLE[normalizeRole(role)] || 'Em Separação - Geral';
}

function orderPayload(order) {
  return {
    id: order.id,
    status_id: order.status_id,
    status_name: order.status_name,
    picked_by: order.picked_by || '',
    picked_by_id: order.picked_by_id || 0,
    picked_from_status_id: order.picked_from_status_id || 0,
    picked_from_status_name: order.picked_from_status_name || '',
  };
}

async function _findStatusByName(name) {
  return db.findOne('base_order_statuses', { name });
}

async function _ensurePersonalStatus(operator) {
  const sid = personalStatusId(Number(operator.id));
  const sname = personalQueueName(operator.name);
  let existing = await db.findOne('base_order_statuses', { id: sid });
  if (existing) {
    if (existing.name !== sname) {
      await db.update('base_order_statuses', { id: sid }, { name: sname });
      existing.name = sname;
    }
    return existing;
  }
  existing = await _findStatusByName(sname);
  if (existing) return existing;

  await db.insert('base_order_statuses', { id: sid, name: sname, color: '#0066FF', count: 0 });
  return db.findOne('base_order_statuses', { id: sid });
}

async function _recalcStatusCounts() {
  const statuses = await db.findMany('base_order_statuses');
  const orders = await db.findMany('base_orders');
  const counts = {};
  for (const o of orders) {
    counts[Number(o.status_id || 0)] = (counts[Number(o.status_id || 0)] || 0) + 1;
  }
  for (const s of statuses) {
    await db.update('base_order_statuses', { id: s.id }, { count: counts[Number(s.id)] || 0 });
  }
}

async function pickupOrder(orderId, operatorId) {
  const op = await db.findOne('base_operators', { id: Number(operatorId) });
  if (!op || !Number(op.is_active)) {
    return { ok: false, error: 'Operador não encontrado ou inativo.' };
  }
  const order = await db.findOne('base_orders', { id: String(orderId) });
  if (!order) return { ok: false, error: 'Pedido não encontrado no banco local.' };

  if ((order.picked_by || '').trim()) {
    return { ok: false, error: `Pedido já está com ${order.picked_by}. Liberar antes de pegar de novo.` };
  }

  const allowed = pickableQueueNames(op.role);
  const current = (order.status_name || '').trim();
  const allowedLower = new Set([...allowed].map((a) => a.toLowerCase()));
  if (!allowedLower.has(current.toLowerCase())) {
    return {
      ok: false,
      error: `Fila '${current}' não é pickable para role ${normalizeRole(op.role)}.`,
    };
  }

  const personal = await _ensurePersonalStatus(op);
  const now = new Date().toISOString();

  await db.update(
    'base_orders',
    { id: order.id },
    {
      picked_from_status_id: Number(order.status_id || 0),
      picked_from_status_name: current,
      picked_by: op.name,
      picked_by_id: Number(op.id),
      picked_at: now,
      status_id: Number(personal.id),
      status_name: personal.name,
    }
  );

  await db.insert(
    'base_order_pickups',
    { order_id: order.id, operator_id: Number(op.id), status_name: personal.name, action: 'pickup', picked_at: now }
  );

  await _recalcStatusCounts();
  const updated = await db.findOne('base_orders', { id: order.id });
  return {
    ok: true,
    action: 'pickup',
    message: `Pedido #${orderId} puxado para ${personal.name}.`,
    order: orderPayload(updated),
  };
}

async function sendOrderToQueue(orderId, operatorId, targetQueue) {
  const op = await db.findOne('base_operators', { id: Number(operatorId) });
  if (!op || !Number(op.is_active)) {
    return { ok: false, error: 'Operador não encontrado ou inativo.' };
  }
  const order = await db.findOne('base_orders', { id: String(orderId) });
  if (!order) return { ok: false, error: 'Pedido não encontrado no banco local.' };

  const role = normalizeRole(op.role);
  const pickedId = Number(order.picked_by_id || 0);
  if (pickedId && pickedId !== Number(op.id) && role !== ROLE_ADMIN) {
    return { ok: false, error: `Pedido está com ${order.picked_by}. Só quem pegou ou Admin pode enviar.` };
  }
  if (!(order.picked_by || '').trim() && !isPersonalStatusName(order.status_name) && !isPersonalStatusId(order.status_id)) {
    return { ok: false, error: 'Pedido não está em fila pessoal. Use Pegar antes de Enviar.' };
  }

  const destName = (targetQueue || '').trim() || defaultSendQueueName(role);
  let dest = await _findStatusByName(destName);
  if (!dest) {
    const maxIdRow = await db.findMany('base_order_statuses', { orderBy: 'id DESC', limit: 1 });
    const newId = (maxIdRow.length ? Number(maxIdRow[0].id) : 0) + 1;
    await db.insert('base_order_statuses',
      { id: newId, name: destName, color: destName.includes('Separ') ? '#b80af7' : '#22a564', count: 0 }
    );
    dest = await _findStatusByName(destName);
  }

  await db.update(
    'base_orders',
    { id: order.id },
    {
      status_id: Number(dest.id),
      status_name: dest.name,
      picked_by: '',
      picked_by_id: 0,
      picked_from_status_id: 0,
      picked_from_status_name: '',
      picked_at: null,
    }
  );

  await db.insert(
    'base_order_pickups',
    { order_id: order.id, operator_id: Number(op.id), status_name: dest.name, action: 'send', picked_at: new Date().toISOString() }
  );

  await _recalcStatusCounts();
  const updated = await db.findOne('base_orders', { id: order.id });
  return {
    ok: true,
    action: 'send',
    message: `Pedido #${orderId} enviado para ${dest.name}.`,
    order: orderPayload(updated),
  };
}

async function releaseOrder(orderId, operatorId) {
  const op = await db.findOne('base_operators', { id: Number(operatorId) });
  if (!op || !Number(op.is_active)) {
    return { ok: false, error: 'Operador não encontrado ou inativo.' };
  }
  const order = await db.findOne('base_orders', { id: String(orderId) });
  if (!order) return { ok: false, error: 'Pedido não encontrado no banco local.' };

  const role = normalizeRole(op.role);
  const pickedId = Number(order.picked_by_id || 0);
  if (pickedId && pickedId !== Number(op.id) && role !== ROLE_ADMIN) {
    return { ok: false, error: `Pedido está com ${order.picked_by}. Só quem pegou ou Admin pode liberar.` };
  }
  if (!(order.picked_by || '').trim() && !isPersonalStatusName(order.status_name)) {
    return { ok: false, error: 'Pedido não está pego (nada a liberar).' };
  }

  let originName = (order.picked_from_status_name || '').trim() || 'Novos pedidos';
  let origin = await _findStatusByName(originName);
  if (!origin && order.picked_from_status_id) {
    origin = await db.findOne('base_order_statuses', { id: Number(order.picked_from_status_id) });
  }
  if (!origin) {
    // Fallback: Novos pedidos
    origin = await _findStatusByName('Novos pedidos');
    if (!origin) {
      // As a last resort, create it
      const maxIdRow = await db.findMany('base_order_statuses', { orderBy: 'id DESC', limit: 1 });
      const newId = (maxIdRow.length ? Number(maxIdRow[0].id) : 0) + 1;
      await db.insert('base_order_statuses', { id: newId, name: 'Novos pedidos', color: '#1e88e5', count: 0 });
      origin = await _findStatusByName('Novos pedidos');
    }
  }

  await db.update(
    'base_orders',
    { id: order.id },
    {
      status_id: Number(origin.id),
      status_name: origin.name,
      picked_by: '',
      picked_by_id: 0,
      picked_from_status_id: 0,
      picked_from_status_name: '',
      picked_at: null,
    }
  );

  await db.insert(
    'base_order_pickups',
    { order_id: order.id, operator_id: Number(op.id), status_name: origin.name, action: 'release', picked_at: new Date().toISOString() }
  );

  await _recalcStatusCounts();
  const updated = await db.findOne('base_orders', { id: order.id });
  return {
    ok: true,
    action: 'release',
    message: `Pedido #${orderId} liberado para ${origin.name}.`,
    order: orderPayload(updated),
  };
}

module.exports = {
  PERSONAL_QUEUE_PREFIX,
  personalQueueName,
  pickupOrder,
  sendOrderToQueue,
  releaseOrder,
  defaultSendQueueName,
  pickableQueueNames,
  isPersonalStatusName,
  isPersonalStatusId,
};

File: apps/api/src/infrastructure/order_pickup.py
"""Pickup local de pedidos: Pegar / Enviar / Liberar (SQLite only)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, Optional, Tuple

from sqlalchemy import select

from src.domain.operator_roles import normalize_role
from src.infrastructure.database import (
    OperatorDB,
    RealOrderDB,
    RealOrderStatusDB,
    async_session,
    init_db,
)
from src.domain.operator_roles import ROLE_ADMIN
from src.infrastructure.native_queues import (
    PERSONAL_STATUS_ID_BASE,
    default_send_queue_name,
    is_personal_status_id,
    is_personal_status_name,
    personal_queue_name,
    personal_status_id,
    pickable_queue_names,
    resolve_status_name_variants,
)

# Re-export para baselinker_status_map e outros
__all__ = [
    "is_personal_status_id",
    "is_personal_status_name",
    "pickup_order",
    "send_order_to_queue",
    "release_order",
]


async def _find_status_by_name(session, name: str) -> Optional[RealOrderStatusDB]:
    for candidate in resolve_status_name_variants(name):
        row = (
            await session.execute(
                select(RealOrderStatusDB).where(RealOrderStatusDB.name == candidate)
            )
        ).scalars().first()
        if row:
            return row
    # case-insensitive fallback
    all_st = (await session.execute(select(RealOrderStatusDB))).scalars().all()
    target = (name or "").strip().lower()
    for s in all_st:
        if (s.name or "").strip().lower() == target:
            return s
        for cand in resolve_status_name_variants(name):
            if (s.name or "").strip().lower() == cand.lower():
                return s
    return None


async def _ensure_personal_status(
    session, operator: OperatorDB
) -> RealOrderStatusDB:
    sid = personal_status_id(int(operator.id))
    sname = personal_queue_name(operator.name)
    existing = await session.get(RealOrderStatusDB, sid)
    if existing:
        if (existing.name or "") != sname:
            existing.name = sname
        return existing
    # Também procura por nome (legado)
    by_name = await _find_status_by_name(session, sname)
    if by_name:
        return by_name
    row = RealOrderStatusDB(id=sid, name=sname, color="#0066FF", count=0)
    session.add(row)
    await session.flush()
    return row


async def _recalc_status_counts(session) -> None:
    statuses = (await session.execute(select(RealOrderStatusDB))).scalars().all()
    orders = (await session.execute(select(RealOrderDB))).scalars().all()
    counts: Dict[int, int] = {}
    for o in orders:
        counts[int(o.status_id or 0)] = counts.get(int(o.status_id or 0), 0) + 1
    for s in statuses:
        s.count = counts.get(int(s.id), 0)


def _order_payload(order: RealOrderDB) -> Dict[str, Any]:
    return {
        "id": order.id,
        "status_id": order.status_id,
        "status_name": order.status_name,
        "picked_by": getattr(order, "picked_by", "") or "",
        "picked_by_id": getattr(order, "picked_by_id", 0) or 0,
        "picked_from_status_id": getattr(order, "picked_from_status_id", 0) or 0,
        "picked_from_status_name": getattr(order, "picked_from_status_name", "") or "",
    }


async def pickup_order(order_id: str, operator_id: int) -> Dict[str, Any]:
    """Puxa pedido da fila geral pickable → Fila · {nome} do operador."""
    await init_db()
    async with async_session() as session:
        operator = await session.get(OperatorDB, int(operator_id))
        if not operator or not operator.is_active:
            return {"ok": False, "error": "Operador não encontrado ou inativo."}

        order = await session.get(RealOrderDB, str(order_id))
        if not order:
            return {"ok": False, "error": "Pedido não encontrado no banco local."}

        already = (getattr(order, "picked_by", None) or "").strip()
        if already:
            return {
                "ok": False,
                "error": f"Pedido já está com {already}. Liberar antes de pegar de novo.",
            }

        role = normalize_role(operator.role)
        allowed = pickable_queue_names(role)
        current = (order.status_name or "").strip()
        if current.lower() not in {a.lower() for a in allowed}:
            return {
                "ok": False,
                "error": (
                    f"Fila '{current}' não é pickable para role {role}. "
                    f"Permitidas: {', '.join(sorted(allowed))}."
                ),
            }

        personal = await _ensure_personal_status(session, operator)
        order.picked_from_status_id = int(order.status_id or 0)
        order.picked_from_status_name = current
        order.picked_by = operator.name
        order.picked_by_id = int(operator.id)
        order.picked_at = datetime.now()
        order.status_id = int(personal.id)
        order.status_name = personal.name

        await _recalc_status_counts(session)
        await session.commit()
        await session.refresh(order)
        return {
            "ok": True,
            "action": "pickup",
            "message": f"Pedido #{order_id} puxado para {personal.name}.",
            "order": _order_payload(order),
        }


async def send_order_to_queue(
    order_id: str,
    operator_id: int,
    target_queue: Optional[str] = None,
) -> Dict[str, Any]:
    """Envia pedido da fila pessoal para destino padrão (ou explícito) por role."""
    await init_db()
    async with async_session() as session:
        operator = await session.get(OperatorDB, int(operator_id))
        if not operator or not operator.is_active:
            return {"ok": False, "error": "Operador não encontrado ou inativo."}

        order = await session.get(RealOrderDB, str(order_id))
        if not order:
            return {"ok": False, "error": "Pedido não encontrado no banco local."}

        role = normalize_role(operator.role)
        picked_id = int(getattr(order, "picked_by_id", 0) or 0)
        if picked_id and picked_id != int(operator.id) and role != ROLE_ADMIN:
            return {
                "ok": False,
                "error": f"Pedido está com {order.picked_by}. Só quem pegou ou Admin pode enviar.",
            }

        if not (getattr(order, "picked_by", None) or "").strip():
            if not is_personal_status_name(order.status_name) and not is_personal_status_id(
                order.status_id
            ):
                return {
                    "ok": False,
                    "error": "Pedido não está em fila pessoal. Use Pegar antes de Enviar.",
                }

        dest_name = (target_queue or "").strip() or default_send_queue_name(role)
        dest = await _find_status_by_name(session, dest_name)
        if not dest:
            max_id = 0
            all_st = (await session.execute(select(RealOrderStatusDB))).scalars().all()
            for s in all_st:
                if int(s.id) < PERSONAL_STATUS_ID_BASE:
                    max_id = max(max_id, int(s.id))
            dest = RealOrderStatusDB(
                id=max_id + 1 if max_id else 50,
                name=dest_name,
                color="#b80af7" if "Separ" in dest_name else "#22a564",
                count=0,
            )
            session.add(dest)
            await session.flush()

        order.status_id = int(dest.id)
        order.status_name = dest.name
        order.picked_by = ""
        order.picked_by_id = 0
        order.picked_from_status_id = 0
        order.picked_from_status_name = ""
        order.picked_at = None

        await _recalc_status_counts(session)
        await session.commit()
        await session.refresh(order)
        return {
            "ok": True,
            "action": "send",
            "message": f"Pedido #{order_id} enviado para {dest.name}.",
            "order": _order_payload(order),
        }


async def release_order(order_id: str, operator_id: int) -> Dict[str, Any]:
    """Libera pedido da fila pessoal → volta à fila geral de origem."""
    await init_db()
    async with async_session() as session:
        operator = await session.get(OperatorDB, int(operator_id))
        if not operator or not operator.is_active:
            return {"ok": False, "error": "Operador não encontrado ou inativo."}

        order = await session.get(RealOrderDB, str(order_id))
        if not order:
            return {"ok": False, "error": "Pedido não encontrado no banco local."}

        picked_id = int(getattr(order, "picked_by_id", 0) or 0)
        role = normalize_role(operator.role)
        if picked_id and picked_id != int(operator.id) and role != "Administrador":
            return {
                "ok": False,
                "error": f"Pedido está com {order.picked_by}. Só quem pegou ou Admin pode liberar.",
            }

        if not (getattr(order, "picked_by", None) or "").strip() and not is_personal_status_name(
            order.status_name
        ):
            return {"ok": False, "error": "Pedido não está pego (nada a liberar)."}

        origin_name = (getattr(order, "picked_from_status_name", None) or "").strip()
        origin_id = int(getattr(order, "picked_from_status_id", 0) or 0)
        origin = None
        if origin_id:
            origin = await session.get(RealOrderStatusDB, origin_id)
        if not origin and origin_name:
            origin = await _find_status_by_name(session, origin_name)
        if not origin:
            # fallback: Novos pedidos / Notebook - Geral
            for fallback in ("Novos pedidos", "Notebook - Geral", "Em Separação - Geral"):
                origin = await _find_status_by_name(session, fallback)
                if origin:
                    break
        if not origin:
            return {"ok": False, "error": "Não foi possível determinar a fila de origem."}

        order.status_id = int(origin.id)
        order.status_name = origin.name
        order.picked_by = ""
        order.picked_by_id = 0
        order.picked_from_status_id = 0
        order.picked_from_status_name = ""
        order.picked_at = None

        await _recalc_status_counts(session)
        await session.commit()
        await session.refresh(order)
        return {
            "ok": True,
            "action": "release",
            "message": f"Pedido #{order_id} liberado para {origin.name}.",
            "order": _order_payload(order),
        }
